File: java/client/src/main/java/glide/api/models/commands/FT/FTInfoOptions.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands.FT;

import static glide.api.models.GlideString.gs;

import glide.api.models.GlideString;
import java.util.ArrayList;

/**
 * Additional arguments for the FT.INFO command.
 *
 * @see <a href="https://valkey.io/commands/ft.info/">valkey.io</a>
 */
public class FTInfoOptions {

    /** Controls which nodes provide index information in cluster mode. */
    public enum InfoScope {
        /** Only the executing (local) node provides index information. This is the default. */
        LOCAL,
        /** Primary nodes of every shard are queried. Only valid in cluster mode. */
        PRIMARY,
        /** All nodes (primary and replica) are queried. Only valid in cluster mode. */
        CLUSTER
    }

    /** Controls shard participation in cluster mode. */
    public enum ShardScope {
        /** Terminate with error if not all shards respond. This is the default. */
        ALLSHARDS,
        /** Generate best-effort reply if not all shards respond within timeout. */
        SOMESHARDS
    }

    /** Controls consistency requirements in cluster mode. */
    public enum ConsistencyMode {
        /** Terminate with error if any response isn't from a consistent index version. Default. */
        CONSISTENT,
        /** Use only responses from nodes with a consistent index version. */
        INCONSISTENT
    }

    private final InfoScope scope;
    private final ShardScope shardScope;
    private final ConsistencyMode consistency;

    public FTInfoOptions(InfoScope scope, ShardScope shardScope, ConsistencyMode consistency) {
        this.scope = scope;
        this.shardScope = shardScope;
        this.consistency = consistency;
    }

    public FTInfoOptions(InfoScope scope) {
        this(scope, null, null);
    }

    public FTInfoOptions(ShardScope shardScope, ConsistencyMode consistency) {
        this(null, shardScope, consistency);
    }

    /** Convert to module API args. */
    public GlideString[] toArgs() {
        ArrayList<GlideString> args = new ArrayList<>();
        if (scope != null) {
            args.add(gs(scope.toString()));
        }
        if (shardScope != null) {
            args.add(gs(shardScope.toString()));
        }
        if (consistency != null) {
            args.add(gs(consistency.toString()));
        }
        return args.toArray(new GlideString[0]);
    }
}

File: go/options/ft_info_options.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package options

// FtInfoScope controls the scope of information returned by FT.INFO.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.info/
type FtInfoScope string

const (
	// FtInfoScopeLocal returns per-node (local shard) index information.
	FtInfoScopeLocal FtInfoScope = "LOCAL"
	// FtInfoScopePrimary returns aggregated information from the primary coordinator.
	// Requires the coordinator module to be enabled.
	FtInfoScopePrimary FtInfoScope = "PRIMARY"
	// FtInfoScopeCluster returns cluster-wide aggregated index information.
	// Requires the coordinator module to be enabled.
	FtInfoScopeCluster FtInfoScope = "CLUSTER"
)

// FtInfoShardScope controls which shards participate in the FT.INFO query.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.info/
type FtInfoShardScope string

const (
	// FtInfoShardScopeAllShards queries all shards (default).
	FtInfoShardScopeAllShards FtInfoShardScope = "ALLSHARDS"
	// FtInfoShardScopeSomeShards queries only a subset of shards.
	FtInfoShardScopeSomeShards FtInfoShardScope = "SOMESHARDS"
)

// FtInfoConsistencyMode controls consistency requirements for the FT.INFO query.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.info/
type FtInfoConsistencyMode string

const (
	// FtInfoConsistencyConsistent requires consistent results across shards.
	FtInfoConsistencyConsistent FtInfoConsistencyMode = "CONSISTENT"
	// FtInfoConsistencyInconsistent allows inconsistent (faster) results.
	FtInfoConsistencyInconsistent FtInfoConsistencyMode = "INCONSISTENT"
)

// FtInfoOptions holds optional arguments for the FT.INFO command.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.info/
type FtInfoOptions struct {
	// Scope controls the scope of information returned.
	Scope FtInfoScope
	// ShardScope controls which shards participate in the query.
	ShardScope FtInfoShardScope
	// Consistency controls consistency requirements.
	Consistency FtInfoConsistencyMode
}

// ToArgs returns the command arguments for FtInfoOptions.
func (o *FtInfoOptions) ToArgs() []string {
	args := []string{}
	if o.Scope != "" {
		args = append(args, string(o.Scope))
	}
	if o.ShardScope != "" {
		args = append(args, string(o.ShardScope))
	}
	if o.Consistency != "" {
		args = append(args, string(o.Consistency))
	}
	return args
}
